### DISCUSSION TITLE
Trigger CanExecute on propertychange of another ViewModel

### DISCUSSION THREAD

**Discussion by Sputnik24** [AvaloniaUI/Avalonia#8000] [Q&A]
Hello,

I'm refering to https://docs.avaloniaui.net/docs/data-binding/binding-to-commands#icommandcanexecute

In principle, this is working for properties of the same viewmodel, but not for properties of other viewmodels.

Background:
I have a MainViewModel of my MainWindow, which is build of several child views using `<UserControl Content="{Binding ChildViewModel}" />` pattern. The child viewmodels are instanciated in the main viewmodel giving them a reference to the main viewmodel in the ctor, so that all childs have access to the public properties of the main viewmodel.

In a child viewmodel, I have something like this:
`[DependsOn(nameof(ChildProperty)),DependsOn(nameof(MainWindowViewModel.MainProperty))]
    public bool CanSend(object parameter) => Main.MainProperty == 1 && ChildProperty == 0;
    public void Send() {}`

Main is the property of ChildViewModel referencing to the MainViewModel instance. Both MainProperty and ChildProperty implement INotifyPropertyChanged:
`private int _mainProperty;
    public int MainProperty {
        get => _mainProperty;
        set => this.RaiseAndSetIfChanged(ref _mainProperty, value);
    }`

If I change the ChildProperty, CanExecute is raised and the button is enabled or disabled. But if I change the MainProperty, nothing happens.

Any idea or hint?

With Prism.Core, everything is working, but I wanted to get rid of Prism.Core using the approach from the documentation.

Code with Prism:
`SendCmd = new DelegateCommand(
                Send,
                () => Main.MainProperty == 1 && ChildProperty == 0)
            .ObservesProperty(() => Main.MainProperty)
            .ObservesProperty(() => ChildProperty);`

**Comment by maxkatz6** [comment]
> DependsOn

No, DependsOn doesn't support CanExecute triggering from changes in another ViewModel.
As a solution, you should be able to create a new intermediate property in the same view model, which contains the command. Then subscribe on PropertyChanged MainWindowViewModel, and call RaiseAndSetIfChanged on this intermediate property. Basically, redirecting changes from another VM to the first one. But...see below.

> I wanted to get rid of Prism.Core using the approach from the documentation.

Documentation contains simple examples to get faster into Avalonia and MVVM. In my opinion, on bigger projects it will be easier to use fully featured MVVM frameworks. It solves quite a lot of boilerplate.

**Comment by Sputnik24** [comment]
@maxkatz6 Thanks for your quick response.

One more questions. I searched the repo for DepensOn and found this LOC: https://github.com/AvaloniaUI/Avalonia/blob/08659fcd7e88df9ec2cf9f85d25f4444b17f31a7/src/Avalonia.Diagnostics/Diagnostics/ViewModels/MainViewModel.cs#L295

Doesn't it exactly, what I want to do?

**Reply by maxkatz6** [reply]
Hm, I actually didn't expect we already support something like this. Seems like so.

**Reply by maxkatz6** [reply]
But something tells me it doesn't actually work
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Data/Converters/MethodToCommandConverter.cs

Can't check it right now.

**Comment by timunie** [comment]
I personally use the MVVM Toolkit for my VM and ICommand. With the latest preview they support also source generators. I can notify CanExecute from the other model if that is needed. 

You can also combine ReactiveUi with this toolkit. So maybe give it a try.

Here is the link https://docs.microsoft.com/en-us/windows/communitytoolkit/mvvm/introduction

Happy coding
Tim

**Comment by Sputnik24** [comment]
@timunie Hi Tim, thank you for your tip. I tried this toolkit, but personally felt like giving too much control away. My goal is to reduce the number of packages and dependencies as I need to publish my project including the full FOSS topic.

I migrated all my DelegateCommands from Prism to ReactiveCommand of ReactiveUI. Makes code look cleaner but also less readable with "WhenAnyValue" instead of "ObservesProperty". Also force me to deal with UIThread.Invoke, but for now it's working.

Nevertheless, the approach shown in the docu with Method() and CanMethod() is nice and clean, less code, don't need to declare all ICommand properties and setting them in the ctor.

**Reply by timunie** [reply]
I love the source gen. You just do this

```
[ICommand]
void DoStuff()
{...}
```

And the rest is done for you. It works with CompiledBindings which is great 👍 and no fancy stuff in the constructor needed. 

https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/3872

But that's just me 2 cents. 😊

**Comment by Sputnik24** [comment]
@timunie I know, I know. With ICommand attribute it's interresting, but would it make also interesting to use ObservableProperty attribute to get rid of all manual RaiseAndSetIfChanged setters. The point when I started to use the MVVM toolkit was, that sometimes I need to do additional stuff in the set of a property and in those cases I could not use ObservableProperty attribute. And again, to reduce number of packages, I decided to stay with ReactiveUI.

But in principle, I like MVVM toolkit, you are right.
